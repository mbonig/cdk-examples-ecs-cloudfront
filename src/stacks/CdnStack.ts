import { Stack, StackProps } from 'aws-cdk-lib';
import { ICertificate } from 'aws-cdk-lib/aws-certificatemanager';
import {
  Distribution,
  OriginProtocolPolicy,
  OriginRequestPolicy,
  ViewerProtocolPolicy,
} from 'aws-cdk-lib/aws-cloudfront';
import { LoadBalancerV2Origin } from 'aws-cdk-lib/aws-cloudfront-origins';
import { ILoadBalancerV2 } from 'aws-cdk-lib/aws-elasticloadbalancingv2';
import { ARecord, HostedZone } from 'aws-cdk-lib/aws-route53';
import { CloudFrontTarget } from 'aws-cdk-lib/aws-route53-targets';
import { Construct } from 'constructs';
import { WebsiteStackTest } from '../constructs/WebsiteStackTest/WebsiteStackTest';

interface CDNStackProps extends StackProps {
  websiteHost: string;
  domainName: string;
  certificate: ICertificate;
  websiteLoadBalancer: ILoadBalancerV2;
}

export class CdnStack extends Stack {
  constructor(scope: Construct, id: string, props: CDNStackProps) {
    super(scope, id, props);
    const websiteFqdn = `${props.websiteHost}.${props.domainName}`;
    const originRequestPolicy = new OriginRequestPolicy(this, 'OriginRequestPolicy', {
      headerBehavior: {
        behavior: 'allViewer',
      },
    });
    const distribution = new Distribution(this, 'Resource', {
      defaultBehavior: {
        origin: new LoadBalancerV2Origin(props.websiteLoadBalancer, {
          protocolPolicy: OriginProtocolPolicy.HTTPS_ONLY,
        }),
        originRequestPolicy: {
          originRequestPolicyId: originRequestPolicy.originRequestPolicyId,
        },
        viewerProtocolPolicy: ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
      },
      certificate: props.certificate,
      domainNames: [websiteFqdn],
    });
    const hostedZone = HostedZone.fromLookup(this, 'HostedZone', { domainName: props.domainName });

    const record = new ARecord(this, 'AliasRecord', {
      recordName: props.websiteHost,
      zone: hostedZone,
      target: {
        aliasTarget: new CloudFrontTarget(distribution),
      },
    });

    const tests = new WebsiteStackTest(this, 'UrlPingTest', {
      url: `https://${websiteFqdn}`,
    });
    tests.node.addDependency(record);

  }

}
