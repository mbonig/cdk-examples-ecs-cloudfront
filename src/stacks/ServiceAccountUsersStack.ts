import { Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';

interface ServiceAccountUsersStackProps extends StackProps {
  readonly serviceAccounts: any[];
}

export class ServiceAccountUsersStack extends Stack {
  constructor(scope: Construct, id: string, props: ServiceAccountUsersStackProps) {
    super(scope, id, props);
  }

}
