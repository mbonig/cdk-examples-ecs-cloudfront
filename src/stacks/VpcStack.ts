import { Stack, StackProps } from 'aws-cdk-lib';
import { IpAddresses, Vpc } from 'aws-cdk-lib/aws-ec2';
import { Construct } from 'constructs';

export interface VpcStackProps extends StackProps {
  cidr: string;
}

export class VpcStack extends Stack {
  public vpc: Vpc;
  constructor(scope: Construct, id: string, props: VpcStackProps) {
    super(scope, id, props);
    this.vpc = new Vpc(this, 'Vpc', {
      maxAzs: 2,
      natGateways: 1,
      ipAddresses: IpAddresses.cidr(props.cidr),
    });
  }

}
