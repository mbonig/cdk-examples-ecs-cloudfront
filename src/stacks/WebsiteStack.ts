import { Duration, Stack, StackProps, Tags } from 'aws-cdk-lib';
import { Certificate, CertificateValidation, ICertificate } from 'aws-cdk-lib/aws-certificatemanager';
import { ITableV2 } from 'aws-cdk-lib/aws-dynamodb';
import { IVpc } from 'aws-cdk-lib/aws-ec2';
import { Platform } from 'aws-cdk-lib/aws-ecr-assets';
import { Cluster, ContainerImage } from 'aws-cdk-lib/aws-ecs';
import {
  ApplicationLoadBalancedFargateService,
  ApplicationLoadBalancedServiceRecordType,
} from 'aws-cdk-lib/aws-ecs-patterns';
import { ILoadBalancerV2 } from 'aws-cdk-lib/aws-elasticloadbalancingv2';
import { HostedZone } from 'aws-cdk-lib/aws-route53';
import { Bucket, IBucket } from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';

export interface WebsiteStackProps extends StackProps {
  tablePrefix: string;
  websiteHost: string;
  domainName: string;
  tables: ITableV2[];
  vpc: IVpc;
}

export class WebsiteStack extends Stack {
  mediaBucket: IBucket;
  loadBalancer: ILoadBalancerV2;
  certificate: ICertificate;

  constructor(scope: Construct, id: string, props: WebsiteStackProps) {
    super(scope, id, props);
    const { vpc, tables } = props;
    const hostedZone = HostedZone.fromLookup(this, 'HostedZone', { domainName: props.domainName });
    const cluster = new Cluster(this, 'Cluster', {
      vpc,
    });
    const mediaBucket = this.mediaBucket = new Bucket(this, 'MediaBucket', {});

    const websiteFqdn = `${props.websiteHost}.${props.domainName}`;

    let certificate = this.certificate = new Certificate(this, 'Certificate', {
      domainName: websiteFqdn,
      validation: CertificateValidation.fromDns(hostedZone),
    });
    const service = new ApplicationLoadBalancedFargateService(this, 'Service', {
      cluster,
      recordType: ApplicationLoadBalancedServiceRecordType.NONE,
      domainName: websiteFqdn,
      domainZone: hostedZone,
      certificate: certificate,
      redirectHTTP: true,
      healthCheckGracePeriod: Duration.minutes(3),
      desiredCount: 2,
      taskImageOptions: {
        image: ContainerImage.fromAsset('website', {
          platform: Platform.custom('linux/amd64'),
        }),
        containerPort: 3000,
        environment: {
          MEDIA_BUCKET: mediaBucket.bucketName,
        },
      },
    });

    this.loadBalancer = service.loadBalancer;

    mediaBucket.grantWrite(service.taskDefinition.taskRole, 'uploads/*');
    mediaBucket.grantRead(service.taskDefinition.taskRole, 'media/*');

    for (const table of tables) {
      table.grantReadWriteData(service.taskDefinition.taskRole);
    }

    Tags.of(this).add('role', 'website');
  }

}
