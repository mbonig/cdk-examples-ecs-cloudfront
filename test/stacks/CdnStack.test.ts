import { App, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { Certificate } from 'aws-cdk-lib/aws-certificatemanager';
import { Vpc } from 'aws-cdk-lib/aws-ec2';
import { ApplicationLoadBalancer } from 'aws-cdk-lib/aws-elasticloadbalancingv2';
import { CdnStack } from '../../src/stacks/CdnStack';

test('CdnStack Snapshot', () => {
  const app = new App();
  const extraStack = new Stack(app, 'Extras', {
    env: {
      account: '000011112222', region: 'us-east-1',
    },
  });
  const domainName = 'example.com';
  const stack = new CdnStack(app, 'test', {
    certificate: new Certificate(extraStack, 'Certificate', {
      domainName: domainName,
    }),
    domainName,
    websiteHost: 'test',
    websiteLoadBalancer: new ApplicationLoadBalancer(extraStack, 'LoadBalancer', {
      vpc: new Vpc(extraStack, 'Vpc'),
    }),
    env: {
      account: '000011112222', region: 'us-east-1',
    },
  });

  const assert = Template.fromStack(stack);
  expect(assert.toJSON()).toMatchSnapshot();
});
