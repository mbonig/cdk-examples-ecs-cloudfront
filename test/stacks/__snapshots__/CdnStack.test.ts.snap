// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CdnStack Snapshot 1`] = `
{
  "Mappings": {
    "AWSCloudFrontPartitionHostedZoneIdMap": {
      "aws": {
        "zoneId": "Z2FDTNDATAQYW2",
      },
      "aws-cn": {
        "zoneId": "Z3RFFRIM2A3IF5",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "AWSCDKTriggerCustomResourceProviderCustomResourceProviderHandler97BECD91": {
      "DependsOn": [
        "AWSCDKTriggerCustomResourceProviderCustomResourceProviderRoleE18FAF0A",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-000011112222-us-east-1",
          "S3Key": "9b642d28372ac0688d94d58f71770ceba997be43db88ea4b1cd0081a3ac5231e.zip",
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "AWSCDKTriggerCustomResourceProviderCustomResourceProviderRoleE18FAF0A",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AWSCDKTriggerCustomResourceProviderCustomResourceProviderRoleE18FAF0A": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "lambda:InvokeFunction",
                  ],
                  "Effect": "Allow",
                  "Resource": [
                    {
                      "Fn::Join": [
                        "",
                        [
                          {
                            "Fn::GetAtt": [
                              "UrlPingTestWebsiteStackTestFunction396DF0FE",
                              "Arn",
                            ],
                          },
                          ":*",
                        ],
                      ],
                    },
                  ],
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "Inline",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "AliasRecord851000D2": {
      "Properties": {
        "AliasTarget": {
          "DNSName": {
            "Fn::GetAtt": [
              "CF9390BA",
              "DomainName",
            ],
          },
          "HostedZoneId": {
            "Fn::FindInMap": [
              "AWSCloudFrontPartitionHostedZoneIdMap",
              {
                "Ref": "AWS::Partition",
              },
              "zoneId",
            ],
          },
        },
        "HostedZoneId": "DUMMY",
        "Name": "test.example.com.",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "CF9390BA": {
      "Properties": {
        "DistributionConfig": {
          "Aliases": [
            "test.example.com",
          ],
          "DefaultCacheBehavior": {
            "CachePolicyId": "658327ea-f89d-4fab-a63d-7e88639e58f6",
            "Compress": true,
            "OriginRequestPolicyId": {
              "Ref": "OriginRequestPolicy3EFDB4FA",
            },
            "TargetOriginId": "testOrigin1194E2589",
            "ViewerProtocolPolicy": "redirect-to-https",
          },
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": [
            {
              "CustomOriginConfig": {
                "OriginProtocolPolicy": "https-only",
                "OriginSSLProtocols": [
                  "TLSv1.2",
                ],
              },
              "DomainName": {
                "Fn::ImportValue": "Extras:ExportsOutputFnGetAttLoadBalancerBE9EEC3ADNSNameD038AF30",
              },
              "Id": "testOrigin1194E2589",
            },
          ],
          "ViewerCertificate": {
            "AcmCertificateArn": {
              "Fn::ImportValue": "Extras:ExportsOutputRefCertificate4E7ABB08F7C8AF50",
            },
            "MinimumProtocolVersion": "TLSv1.2_2021",
            "SslSupportMethod": "sni-only",
          },
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "OriginRequestPolicy3EFDB4FA": {
      "Properties": {
        "OriginRequestPolicyConfig": {
          "CookiesConfig": {
            "CookieBehavior": "none",
          },
          "HeadersConfig": {
            "HeaderBehavior": "allViewer",
          },
          "Name": "testOriginRequestPolicy621CF7B0",
          "QueryStringsConfig": {
            "QueryStringBehavior": "none",
          },
        },
      },
      "Type": "AWS::CloudFront::OriginRequestPolicy",
    },
    "UrlPingTestE1169BF6": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "AliasRecord851000D2",
      ],
      "Properties": {
        "ExecuteOnHandlerChange": true,
        "HandlerArn": {
          "Ref": "UrlPingTestWebsiteStackTestFunctionCurrentVersionB2F14C763c8f9572ded751b5c7825e5cc99d11e4",
        },
        "InvocationType": "RequestResponse",
        "ServiceToken": {
          "Fn::GetAtt": [
            "AWSCDKTriggerCustomResourceProviderCustomResourceProviderHandler97BECD91",
            "Arn",
          ],
        },
        "Timeout": "600000",
      },
      "Type": "Custom::Trigger",
      "UpdateReplacePolicy": "Delete",
    },
    "UrlPingTestWebsiteStackTestFunction396DF0FE": {
      "DependsOn": [
        "AliasRecord851000D2",
        "UrlPingTestWebsiteStackTestFunctionServiceRole15EF1E4A",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-000011112222-us-east-1",
          "S3Key": "a886fb33721bee5d86d33990a15564ae555ffa2cdeb2d17fbb6227adaa068800.zip",
        },
        "Description": "src/constructs/WebsiteStackTest/WebsiteStackTest.lambda.ts",
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "URL": "https://test.example.com",
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "UrlPingTestWebsiteStackTestFunctionServiceRole15EF1E4A",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "UrlPingTestWebsiteStackTestFunctionCurrentVersionB2F14C763c8f9572ded751b5c7825e5cc99d11e4": {
      "DependsOn": [
        "AliasRecord851000D2",
      ],
      "Properties": {
        "FunctionName": {
          "Ref": "UrlPingTestWebsiteStackTestFunction396DF0FE",
        },
      },
      "Type": "AWS::Lambda::Version",
    },
    "UrlPingTestWebsiteStackTestFunctionServiceRole15EF1E4A": {
      "DependsOn": [
        "AliasRecord851000D2",
      ],
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
